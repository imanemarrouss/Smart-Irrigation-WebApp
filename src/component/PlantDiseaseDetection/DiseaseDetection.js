import React, { useState } from "react";
import styled from "styled-components";
import Button from "./Button";
import Form from "./Form";

const DiseaseDetection = () => {
  const [image, setImage] = useState(null); // Store the uploaded image
  const [result, setResult] = useState(""); // Store the detection result

  // Function to handle image upload
  const handleImageChange = (e) => {
    setImage(e.target.files[0]); // Store the uploaded file
  };

  // Function to handle form submission
  const handleSubmit = async (e) => {
    e.preventDefault();

    if (!image) {
      alert("Please upload an image.");
      return;
    }

    // Create a form data object to send the image to the backend
    const formData = new FormData();
    formData.append("image", image);

    try {
      // Make the API call to Flask backend
      const response = await fetch("http://localhost:3000/detect", {
        // Adjust the URL based on your Flask server setup
        method: "POST",
        body: formData,
      });

      const data = await response.json();
      setResult(data.result); // Assuming the result is returned in this format
    } catch (error) {
      console.error("Error uploading image:", error);
    }
  };

  return (
    <StyledContainer>
      <div>
        <Title data-testid="title" >Check Your Plant Health</Title> {/* Add a title here */}
        <Form handleImageChange={handleImageChange} />{" "}
        {/* Pass the handler to Form */}
        <Button testId={"submit-button-disease-detection"} children=""  onClick={handleSubmit} /> {/* Trigger the form submission */}
        {result && <Result>{result}</Result>}{" "}
        {/* Display the result if available */}
      </div>
    </StyledContainer>
  );
};

// Styled components for title and result display
const StyledContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  block-size: 100vh;
  background-image: url("/path/to/your/background.jpg");
  background-size: cover;
  background-position: center;
`;

const Title = styled.h1`
  color: #2e8b57; /* Green color */
  font-size: 48px;
  margin-block-end: 20px;
  text-align: center;
`;

const Result = styled.p`
  margin-block-start: 20px;
  font-size: 24px;
  color: #333;
  text-align: center;
`;

export default DiseaseDetection;