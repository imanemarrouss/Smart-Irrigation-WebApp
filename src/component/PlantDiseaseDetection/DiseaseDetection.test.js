/*import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import DiseaseDetection from './DiseaseDetection';

global.fetch = jest.fn();

describe('DiseaseDetection Component', () => {
  beforeEach(() => {
    fetch.mockClear();
  });

  test('renders the component correctly', () => {
    render(<DiseaseDetection />);
    
    expect(screen.getByTestId('title')).toHaveTextContent('Check Your Plant Health');
    expect(screen.getByTestId('submit-button-disease-detection')).toBeInTheDocument();
  });

  test('allows image upload', () => {
    render(<DiseaseDetection />);
    
    const file = new File(['test'], 'test.png', { type: 'image/png' });
    const fileInput = screen.getByTestId('form').querySelector('input[type="file"]');
    fireEvent.change(fileInput, { target: { files: [file] } });

    expect(fileInput.files[0]).toBe(file);
    expect(fileInput.files[0].name).toBe('test.png');
  });


  /* error 
  test('handles API error gracefully', async () => {
    const consoleErrorMock = jest.spyOn(console, 'error').mockImplementation(() => {});
    fetch.mockRejectedValueOnce(new Error('Network Error'));

    render(<DiseaseDetection />);
    
    const file = new File(['test'], 'test.png', { type: 'image/png' });
    const fileInput = screen.getByTestId('form').querySelector('input[type="file"]');
    fireEvent.change(fileInput, { target: { files: [file] } });

    const submitButton = screen.getByTestId('submit-button-disease-detection');
    fireEvent.click(submitButton);

    await waitFor(() => {
      expect(fetch).toHaveBeenCalledWith('http://localhost:5000/detect', {
        method: 'POST',
        body: expect.any(FormData),
      });
    });

    consoleErrorMock.mockRestore();
  });
}); 

 //error with mock
 /*test("shows alert when submitting without an image", () => {
    // Create a spy on window.alert
    const mockMethod = jest.fn();
    alertMock = jest.spyOn(window, 'alert').mockImplementation(() => {
        return {
          method: mockMethod,
        };
      });
 
    // Render the component
    render(<DiseaseDetection />);
 
    // Find the submit button
    const submitButton = screen.getByTestId('submit-button-disease-detection');
    
    // Log additional debugging information
    console.log('Submit button exists:', !!submitButton);
    console.log('Button properties:', {
        testId: submitButton.getAttribute('data-testid'),
        type: submitButton.type,
        disabled: submitButton.disabled
    });

    // Try to simulate click
    try {
        fireEvent.click(submitButton);
        console.log('Click event fired successfully');
    } catch (error) {
        console.error('Error firing click event:', error);
    }
   
    // Check alert mock
    console.log('Alert mock calls:', alertMock.mock.calls);
 
    // Assertions
    expect(alertMock).toHaveBeenCalledTimes(1);
    expect(alertMock).toHaveBeenCalledWith("Please upload an image.");
 
    // Restore the original alert function
    alertMock.mockRestore();
});
*/

  
//});

/*test('successfully detects disease with API response', async () => {
    const mockResponse = {
      ok: true,
      json: () => Promise.resolve({ result: 'Healthy Plant' }),
    };
    fetch.mockResolvedValueOnce(mockResponse);

    render(<DiseaseDetection />);
    
    const file = new File(['test'], 'test.png', { type: 'image/png' });
    const fileInput = screen.getByTestId('form').querySelector('input[type="file"]');
    fireEvent.change(fileInput, { target: { files: [file] } });

    const submitButton = screen.getByTestId('submit-button-disease-detection');
    fireEvent.click(submitButton);

    await waitFor(() => {
      expect(fetch).toHaveBeenCalledWith('http://localhost:5000/detect', {
        method: 'POST',
        body: expect.any(FormData),
      });
    });

    await waitFor(() => {
      expect(screen.getByTestId('result')).toHaveTextContent('Healthy Plant');
    });
 });*/

 import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import DiseaseDetection from './DiseaseDetection';

// Mock the child components
jest.mock('./Button', () => {
  return function MockButton({ onClick }) {
    return <button onClick={onClick}>Submit</button>;
  };
});

jest.mock('./Form', () => {
  return function MockForm({ handleImageChange }) {
    return (
      <input
        type="file"
        data-testid="file-input"
        onChange={handleImageChange}
      />
    );
  };
});

// Mock the fetch function
global.fetch = jest.fn();

describe('DiseaseDetection Component', () => {
  beforeEach(() => {
    // Clear all mocks before each test
    jest.clearAllMocks();
  });

  test('renders correctly with initial state', () => {
    render(<DiseaseDetection />);
    
    // Check if the title is rendered
    expect(screen.getByText('Check Your Plant Health')).toBeInTheDocument();
    
    // Check if form elements are present
    expect(screen.getByTestId('file-input')).toBeInTheDocument();
    expect(screen.getByText('Submit')).toBeInTheDocument();
    
    // Result should not be present initially
    expect(screen.queryByText(/disease detected/i)).not.toBeInTheDocument();
  });

  test('handles image upload', () => {
    render(<DiseaseDetection />);
    
    const file = new File(['dummy content'], 'test.png', { type: 'image/png' });
    const fileInput = screen.getByTestId('file-input');
    
    fireEvent.change(fileInput, { target: { files: [file] } });
    
    // The image should be stored in the component's state
    // (we can't directly test state, but we can test the effects)
  });

  test('shows alert when submitting without image', () => {
    const alertMock = jest.spyOn(window, 'alert').mockImplementation(() => {});
    render(<DiseaseDetection />);
    
    const submitButton = screen.getByText('Submit');
    fireEvent.click(submitButton);
    
    expect(alertMock).toHaveBeenCalledWith('Please upload an image.');
    alertMock.mockRestore();
  });

  test('handles successful API call', async () => {
    const mockResponse = { result: 'Healthy plant detected' };
    fetch.mockImplementationOnce(() =>
      Promise.resolve({
        json: () => Promise.resolve(mockResponse),
      })
    );

    render(<DiseaseDetection />);
    
    // Upload a file
    const file = new File(['dummy content'], 'test.png', { type: 'image/png' });
    const fileInput = screen.getByTestId('file-input');
    fireEvent.change(fileInput, { target: { files: [file] } });
    
    // Submit the form
    const submitButton = screen.getByText('Submit');
    fireEvent.click(submitButton);
    
    // Wait for the result to be displayed
    await waitFor(() => {
      expect(screen.getByText(mockResponse.result)).toBeInTheDocument();
    });
    
    // Verify API call
    expect(fetch).toHaveBeenCalledTimes(1);
    expect(fetch).toHaveBeenCalledWith('http://localhost:3000/detect', {
      method: 'POST',
      body: expect.any(FormData),
    });
  });

  test('handles API error', async () => {
    // Mock console.error to prevent error output in test logs
    const consoleSpy = jest.spyOn(console, 'error').mockImplementation(() => {});
    
    fetch.mockImplementationOnce(() =>
      Promise.reject(new Error('API Error'))
    );

    render(<DiseaseDetection />);
    
    // Upload a file
    const file = new File(['dummy content'], 'test.png', { type: 'image/png' });
    const fileInput = screen.getByTestId('file-input');
    fireEvent.change(fileInput, { target: { files: [file] } });
    
    // Submit the form
    const submitButton = screen.getByText('Submit');
    fireEvent.click(submitButton);
    
    // Wait for error to be logged
    await waitFor(() => {
      expect(consoleSpy).toHaveBeenCalledWith(
        'Error uploading image:',
        expect.any(Error)
      );
    });
    
    consoleSpy.mockRestore();
  });

  test('matches snapshot', () => {
    const { container } = render(<DiseaseDetection />);
    expect(container).toMatchSnapshot();
  });
});