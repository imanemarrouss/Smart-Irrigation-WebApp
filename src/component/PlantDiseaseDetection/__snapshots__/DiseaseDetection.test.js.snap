// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiseaseDetection Component matches snapshot 1`] = `
<div>
  <div
    class="sc-blHHSb coFXrE"
  >
    <div>
      <h1
        class="sc-gtLWhw hRBIIe"
      >
        Check Your Plant Health
      </h1>
       
      <input
        data-testid="file-input"
        type="file"
      />
       
      <button>
        Submit
      </button>
       
       
    </div>
  </div>
</div>
`;
